src/storage/src/hummock/store/version.rs

```diff
@@ -235,3 +264,364 @@ impl HummockReadVersion {
         self.staging.sst.clear();
     }
 }
+
+pub fn read_filter_for_batch(
+    epoch: HummockEpoch, // for check
+    table_id: TableId,
+    key_range: &(Bound<Vec<u8>>, Bound<Vec<u8>>),
+    read_version_vec: Vec<Arc<RwLock<HummockReadVersion>>>,
+) -> StorageResult<(Vec<ImmutableMemtable>, Vec<SstableInfo>, CommittedVersion)> {
+    assert!(!read_version_vec.is_empty());
+    let read_version_guard_vec = read_version_vec
+        .iter()
+        .map(|read_version| read_version.read())
+        .collect_vec();
+    let mut imm_vec = Vec::default();
+    let mut sst_vec = Vec::default();
+    // to get max_mce with lock_guard to avoid losing committed_data since the read_version
+    // update is asynchronous
+    let (lastst_committed_version, max_mce) = {
+        let committed_version = read_version_guard_vec
+            .iter()
+            .max_by_key(|read_version| read_version.committed().max_committed_epoch())
+            .unwrap()
+            .committed();
+
+        (
+            committed_version.clone(),
+            committed_version.max_committed_epoch(),
+        )
+    };
+
+    // only filter the staging data that epoch greater than max_mce to avoid data duplication
+    let (min_epoch, max_epoch) = (max_mce, epoch);
+
+    // prune imm and sst with max_mce
+    for read_version_guard in read_version_guard_vec {
+        let (imm_iter, sst_iter) = read_version_guard
+            .staging()
+            .prune_overlap(min_epoch, max_epoch, table_id, key_range);
```

Comment:
The sstableinfo of each `read_version` may be repeated.

Reply:
Thanks , before introduce new uploader, the staging sst will not be updated, so in this PR , add TODO coments first and dedup sstable_infos in next PR